File: trees/ZigzagLevelOrder.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
from queue import Queue

class Solution:
    def zigzagLevelOrder(self, root: Optional[TreeNode]) -> List[List[int]]:
        ordered = []
        # return an empty list if root is null
        if root is None: return ordered
        queue = Queue()
        # add root
        queue.put(root)
        zigzag = False
        while (not queue.empty()):
            level = []
            q_size = queue.qsize()
            for i in range (q_size):
                node = queue.get()
                if zigzag:
                    level.insert(0, node.val)
                else:
                    level.append(node.val)
                # add left node and right node if not null
                if (node.left != None):
                    queue.put(node.left)
                if(node.right != None):
                    queue.put(node.right)
            # add this level
            ordered.append(level)
            zigzag = not zigzag
        return ordered

File: trees/ZigzagLevelOrder.cs
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     public int val;
 *     public TreeNode left;
 *     public TreeNode right;
 *     public TreeNode(int val=0, TreeNode left=null, TreeNode right=null) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
public class Solution {
    public IList<IList<int>> ZigzagLevelOrder(TreeNode root) {
        IList<IList<int>> order = new List<IList<int>>();
        // return an empty list if root is null
        if (root == null) return order;
        Queue<TreeNode> queue = new Queue<TreeNode>();
        // add root
        queue.Enqueue(root);
        bool zigzag = false;
        while(queue.Any()){
            List<int> level = new List<int>();
            int q_size = queue.Count();
            for (int i=0; i< q_size; i++){
                TreeNode node = queue.Dequeue();
                if(zigzag){
                    level.Insert(0, node.val);
                } else {
                    level.Add(node.val);
                }
                // add left node and right node if not null
                if(node.left != null){
                    queue.Enqueue(node.left);
                } 
                if(node.right !=null){
                    queue.Enqueue(node.right);
                }
            }
            order.Add(level);
            zigzag = !zigzag;
        }
        return order;        
    }
}